aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -196,7 +197,29 @@ private void openStream() throws IOException {
 
   private void closeStream() throws IOException {
     if (stream != null) {
-      stream.close();
+      // if we aren't at the end of the stream, and the stream is abortable, then
+      // call abort() so we don't read the remaining data with the Apache HTTP client
+      abortStream();
+      try {
+        stream.close();
+      } catch (IOException e) {
+        // the Apache HTTP client will throw a ConnectionClosedException
+        // when closing an aborted stream, which is expected
+        if (!e.getClass().getSimpleName().equals("ConnectionClosedException")) {
```

Comment:
Can we just do `instanceof` here?

Reply:
Oh okay I see this thread https://github.com/apache/iceberg/pull/7262#discussion_r1156495423, since it's specific to Apache Http client, it's not guaranteed it'll be on the classpath so we can't use instanceof

Reply:
That would add a dependency on the Apache HTTP client. Also, the AWS bundle shades the library, so the classes will be different depending on that (different packages).